### PR TITLE
[16.0][MIG] stock_picking_analytic: Migration to 16.0

### PR DESCRIPTION
**Pull request by yostashiro**
* Switch field from `analytic_account_id` to `analytic_distribution`.
* ~~Change depends referenced field from `move_ids_without_package` to `move_ids`.  `move_ids_without_package` is a non-stored compute field and therefore is not appropriate to be used as a trigger (you get a warning message in running tests without this change).~~ Add dummy search arg to `move_ids_without_package` to bypass warning.
* Remove `groups="analytic.group_analytic_accounting"` from the `analytic_distribution` field added in the picking form, because it would cause an error `Field 'analytic_distribution' used in context ({'default_company_id': company_id, 'default_date': scheduled_date, 'default_date_deadline': date_deadline, 'picking_type_code': picking_type_code, 'default_picking_id': id, 'form_view_ref': 'stock.view_move_form', 'address_in_id': partner_id, 'default_picking_type_id': picking_type_id, 'default_location_id': location_id, 'default_location_dest_id': location_dest_id, 'default_partner_id': partner_id, 'default_analytic_distribution': analytic_distribution}) is restricted to the group(s) analytic.group_analytic_accounting` in loading the view.  Looks like this refactoring https://github.com/odoo/odoo/commit/0501bbd62e517f6c215d9e7e36d61747c7f5816b has added the check.  Not sure there is a good way to still add the group to the field, however there is not much point doint it since the module wouldn't be installed without enabling the analytic accounting.  -> Add an invisible `analytic_distribution` field without group assignment (thanks @luisg123v ).
* ~~Extension of the search view to add `analytic_distribution` field is disabled as the search on json fields do not work in a meaningful manner as of now.~~

@qrtl

### CODE REVIEW

**Comment by rousseldenis** [comment]
@yostashiro Shouldn't this deserves a migration script ?

**Comment by yostashiro** [comment]
> @yostashiro Shouldn't this deserves a migration script ?

@rousseldenis I was thinking no, as I thought the whole point of having analytic account (analytic distribution now) in the picking was to propagate the value to stock moves, and few people would care about the data persistency of itself (but I might be wrong).  If anything, migration script should be added in https://github.com/OCA/account-analytic/tree/16.0/stock_analytic for analytic distribution of stock moves?

**Review by desdelinux — APPROVED** [review]
Functional review is ok.

cc @rolandojduartem @rousseldenis @luisg123v

**Comment by desdelinux** [comment]
> > @yostashiro Shouldn't this deserves a migration script ?
> 
> @rousseldenis I was thinking no, as I thought the whole point of having analytic account (analytic distribution now) in the picking was to propagate the value to stock moves, and few people would care about the data persistency of itself (but I might be wrong). If anything, migration script should be added in https://github.com/OCA/account-analytic/tree/16.0/stock_analytic for analytic distribution of stock moves?

@rousseldenis I agree with this

**Comment by rolandojduartem** [comment]
LGTM +1

**Review by rolandojduartem — APPROVED** [review]

File: stock_picking_analytic/models/stock_picking.py
Comment (luisg123v) [inline stock_picking_analytic/models/stock_picking.py]: If you use a shorter hash, line won't be as long and noqa won't be required, e.g.:

```suggestion
    # https://github.com/odoo/odoo/blob/ae94f14a/odoo/fields.py#L808-L812
```

File: stock_picking_analytic/tests/test_picking_analytic_account.py
Comment (luisg123v) [inline stock_picking_analytic/tests/test_picking_analytic_account.py]: Prefer the Command syntax

Same below

File: stock_picking_analytic/tests/test_picking_analytic_account.py
Comment (luisg123v) [inline stock_picking_analytic/tests/test_picking_analytic_account.py]:
```suggestion
        self.assertTrue(picking.move_ids)
```

Same below

File: stock_picking_analytic/views/stock_picking.xml
Comment (luisg123v) [inline stock_picking_analytic/views/stock_picking.xml]: You just need to add the field again as invisible, so the field is included no matter if current user belongs to the group

**Review by luisg123v — CHANGES_REQUESTED** [review]
You're including commits that are not merged in the stable branch, some of which are even WIP:
- 452e4260144a3a6f510a6004f6ce86b2e7be156e
- 74e5d890f0498717a14fdb54443a5a18eaf8b83e

Those are migration commits to 15.0, but this module is not migrated to 15.0 yet. I'd suggest either:
- removing those ones from this migration
- Or  creating (first) a migration PR to 15.0 and, after that one is merged, resume this one

On the other hand:

"pre-commit" -> "pre-commit stuff"

**Comment by luisg123v** [comment]
PR description is a comprehensive one, it would be a shame if it's not included also in commit description.

**Comment by desdelinux** [comment]
@yostashiro Hello, you can apply the suggested changes

**Comment by yostashiro** [comment]
@luisg123v @desdelinux Thanks for your review comments. Will try and follow up on them but give me some time. A bit swamped at the moment.

**Comment by rousseldenis** [comment]
@yostashiro Any news on this ?

**Comment by yostashiro** [comment]
I plan to update the PR this weekend.

**Comment by yostashiro** [comment]
@luisg123v @desdelinux @rousseldenis Thanks for waiting. Just updated the PR. I believe I took care of all the points in the feedback.

**Review by desdelinux — APPROVED** [review]
Functional Review is ok.

cc @luisg123v @rousseldenis

**Comment by rolandojduartem** [comment]
@luisg123v could you review this PR?

File: stock_picking_analytic/views/stock_picking.xml
Comment (luisg123v) [inline stock_picking_analytic/views/stock_picking.xml]: Typo:  assigining -> assigning

But I think this comment should be removed, this is not a special case but just the way v16 works. Most of views containing fields with groups (e.g. almost all views with the `company_id` field) define the same field as invisible.

A similar case occurred in previous versions when using a field in attrs that was not already present in the view

File: stock_picking_analytic/views/stock_picking.xml
Comment (luisg123v) [inline stock_picking_analytic/views/stock_picking.xml]: Since it's invisible, widget is meaningless

File: stock_picking_analytic/tests/test_picking_analytic_account.py
Comment (luisg123v) [inline stock_picking_analytic/tests/test_picking_analytic_account.py]: Why update instead of write?

write is preferred, because update writes fields one by one

Same below

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: stock_picking_analytic/tests/test_picking_analytic_account.py
Comment (yostashiro) [reply stock_picking_analytic/tests/test_picking_analytic_account.py]: `write()` would throw an error for `picking.move_ids_without_package` not having the database record when tests started.  I have switched to assign `move_ids` instead of `move_ids_without_package` to the picking in the setup to use `write()` in tests.

**Review by desdelinux — APPROVED** [review]
Functional Review is ok.

@rousseldenis

**Comment by desdelinux** [comment]
@OCA/accounting-maintainers This one is ready

**Review by desdelinux — APPROVED** [review]

**Review by rolandojduartem — APPROVED** [review]

**Review by luisg123v — APPROVED** [review]
LGTM :+1:

**Comment by luisg123v** [comment]
@moylop260,

Could you review/merge, please?

Regards,

**Pull request merged**
